ASoC: intel: sof_rt5682 depends on SND_SOC_SOF_HDA_LINK

libinyang opened the pull request:
SOF audio driver uses SND_SOC_SOF_HDA_LINK to control whether
HDA link codec driver can be enabled or not.

User will be confused if SND_SOC_SOF_HDA_LINK is disabled but
hdac_hdmi driver is still enabled for SOF audio driver.

Signed-off-by: Libin Yang <libin.yang@linux.intel.com>

plbossart commented on sound/soc/intel/boards/Kconfig:
```diff
@@ -454,7 +454,7 @@ config SND_SOC_INTEL_SOF_CML_RT1011_RT5682_MACH
         select SND_SOC_RT5682
         select SND_SOC_DMIC
-        select SND_SOC_HDAC_HDMI
+        select SND_SOC_HDAC_HDMI if SND_SOC_SOF_HDA_LINK
```
this one makes no sense, see line 442, you are already in the case where HDA_LINK is required.

libinyang replied:
@plbossart you are right.

plbossart commented on sound/soc/intel/boards/Kconfig:
```diff
@@ -431,7 +431,7 @@ config SND_SOC_INTEL_SOF_RT5682_MACH
 	select SND_SOC_RT5682
 	select SND_SOC_DMIC
-	select SND_SOC_HDAC_HDMI
+	select SND_SOC_HDAC_HDMI if SND_SOC_SOF_HDA_LINK
```
this one makes no sense, or the fix is in the wrong position. The code in the machine drivers does not have conditional compilation support, so if you don't have SOF_HDA_LINK things will be broken.

I think the right fix would be to change line 426 as well:
if SND_SOC_SOF_BAYTRAIL || SND_SOC_SOF_HDA_LINK

@bardliao what do you think?

bardliao replied:
Agree with @plbossart  We should always select SND_SOC_HDAC_HDMI if we want to use sof_rt5682 machine driver.

libinyang replied:
@plbossart OK. I will make the patch as what you suggested.

libinyang commented:
> Looks good, but the "user will be confused" got me a bit. What actually happens?
> 
> This patch will conflict with my snd-hda-codec-hdmi series in upstream. Not a blocker, but something to note for maintainers.

@kv2019i Thanks for review. Sometimes, user may want to disable HDMI audio only. Let's consider such situation, there are many machine drivers are selected, some machine driver will always select hdac_hdmi audio. User won't to disable hdac_hdmi audio on his platform. Mostly he will disable SND_SOC_SOF_HDA_LINK instead of disable all unrelated machine drivers which will enable hdac_hdmi. In this case, although he disables SND_SOC_SOF_HDA_LINK, hdac_hdmi audio driver will also be enabled by other machine drivers.

If I changed the patch as Pierre's suggestion, will it conflict with you patch?

libinyang commented:
> 
> If I changed the patch as Pierre's suggestion, will it conflict with you patch?

@kv2019i I have updated the patch based on Pierre's comments. Please help check whether it is comfortable to you.

plbossart commented on sound/soc/intel/boards/Kconfig:
```diff
@@ -423,7 +423,7 @@ config SND_SOC_INTEL_SKL_HDA_DSP_GENERIC_MACH
 endif ## SND_SOC_INTEL_SKYLAKE_HDAUDIO_CODEC || SND_SOC_SOF_HDA_AUDIO_CODEC
 
-if SND_SOC_SOF_HDA_COMMON || SND_SOC_SOF_BAYTRAIL
+if (SND_SOC_SOF_HDA_COMMON || SND_SOC_SOF_BAYTRAIL) && SND_SOC_SOF_HDA_LINK
```
the logic seems wrong.
If you only want to compile for Baytrail, this will prevent this machine driver from being selected.

plbossart replied:
like I said earlier, you should have 

````
if (SND_SOC_SOF_HDA_LINK || SND_SOC_SOF_BAYTRAIL)
    select HDAC_HDMI if SND_SOC_SOF_HDA_LINK
````

libinyang replied:
@plbossart 

```
select HDAC_HDMI if SND_SOC_SOF_HDA_LINK
```
should be 
```
select HDAC_HDMI
```
as we have judged
```
 if (SND_SOC_SOF_HDA_LINK || SND_SOC_SOF_BAYTRAIL)
```
right?

Do you mean we should 
```
if SND_SOC_SOF_BAYTRAIL || SND_SOC_SOF_HDA_LINK
config SND_SOC_INTEL_SOF_RT5682_MACH
	tristate "SOF with rt5682 codec in I2S Mode"
	depends on I2C && ACPI
	depends on (SND_SOC_SOF_HDA_COMMON && (MFD_INTEL_LPSS || COMPILE_TEST)) ||\
		   (SND_SOC_SOF_BAYTRAIL && (X86_INTEL_LPSS || COMPILE_TEST))
	select SND_SOC_RT5682
	select SND_SOC_DMIC
	select SND_SOC_HDAC_HDMI
	help
	   This adds support for ASoC machine driver for SOF platforms
	   with rt5682 codec.
	   Say Y if you have such a device.
	   If unsure select "N".
endif
```

kv2019i commented:
@libinyang wrote:
> @kv2019i Thanks for review. Sometimes, user may want to disable HDMI audio only. Let's consider such situation, there are many machine drivers are selected, some machine driver will always select hdac_hdmi audio. User won't to disable hdac_hdmi audio on his platform. Mostly he will disable SND_SOC_SOF_HDA_LINK instead of disable all unrelated machine drivers which will enable hdac_hdmi. In this case, although he disables SND_SOC_SOF_HDA_LINK, hdac_hdmi audio driver will also be enabled by other machine drivers.

So the confusion is about SND_SOC_SOF_HDA_LINK=0 not having the expected affected on the rt5682 driver. Ok, thanks, now I understand. 

> If I changed the patch as Pierre's suggestion, will it conflict with you patch?

It will conflict still, but don't worry about that. With long review rounds like this one for my series, conflicts are bound to happen.

plbossart commented:
@libinyang no, you only select HDAC_HDMI if you have an HDaudio platform with the HDA_LINK enabled. You don't select it for Baytrail only.

libinyang commented:
> @libinyang no, you only select HDAC_HDMI if you have an HDaudio platform with the HDA_LINK enabled. You don't select it for Baytrail only.

@plbossart I looked over the discussion in https://github.com/thesofproject/linux/pull/1241 and talked with @bardliao these days. I found it's very hard to work out a perfect solution. I need some days to think it over. Thanks for the information.

libinyang commented:
@plbossart @bardliao @kv2019i 
Sorry for pending this topic for a long time. Finally I'm back to this issue now. Last 2 weeks I was on business travel for EHL PO. 

Here is my some thoughts on this issue.

In our kconfig, if you select some machine drivers, it will automatically select hdmi audio driver. However, I don't think this is a good action. The machine driver should always let user select whether to support hdmi audio or not. This is because user may use external graphics cards. This requires to disable hdmi audio in sof. Otherwise some issues may happen.

So my suggestion is:
1. In every machine driver, we should support both mode (with hdmi or without hdmi)
2. we will supply 2 kinds of topology (with hdmi or without hdmi)
3. The driver will automatically load the correct topology based on hdmi being enabled or disabled.

What's your opinion?

plbossart commented:
> @plbossart @bardliao @kv2019i
> Sorry for pending this topic for a long time. Finally I'm back to this issue now. Last 2 weeks I was on business travel for EHL PO.
> 
> Here is my some thoughts on this issue.
> 
> In our kconfig, if you select some machine drivers, it will automatically select hdmi audio driver. However, I don't think this is a good action. The machine driver should always let user select whether to support hdmi audio or not. This is because user may use external graphics cards. This requires to disable hdmi audio in sof. Otherwise some issues may happen.
> 
> So my suggestion is:
> 
> 1. In every machine driver, we should support both mode (with hdmi or without hdmi)
> 2. we will supply 2 kinds of topology (with hdmi or without hdmi)
> 3. The driver will automatically load the correct topology based on hdmi being enabled or disabled.
> 
> What's your opinion?

we will move to support multiple clients and all machine drivers will no longer support HDMI, we will have a dedicated HDMI card. problem solved.

libinyang commented:
> > @plbossart @bardliao @kv2019i
> > Sorry for pending this topic for a long time. Finally I'm back to this issue now. Last 2 weeks I was on business travel for EHL PO.
> > Here is my some thoughts on this issue.
> > In our kconfig, if you select some machine drivers, it will automatically select hdmi audio driver. However, I don't think this is a good action. The machine driver should always let user select whether to support hdmi audio or not. This is because user may use external graphics cards. This requires to disable hdmi audio in sof. Otherwise some issues may happen.
> > So my suggestion is:
> > 
> > 1. In every machine driver, we should support both mode (with hdmi or without hdmi)
> > 2. we will supply 2 kinds of topology (with hdmi or without hdmi)
> > 3. The driver will automatically load the correct topology based on hdmi being enabled or disabled.
> > 
> > What's your opinion?
> 
> we will move to support multiple clients and all machine drivers will no longer support HDMI, we will have a dedicated HDMI card. problem solved.

@plbossart Got it. Thanks.

libinyang commented:
As @plbossart said, after moving to multiple clients support, this will never be a problem. So let's close it.
